README
README for MCPrimers.pm V2.1, mcprimers.pl, and CloningVector.pm

1. Authorship and copyright

Author:    Stephen G. Lenk, November 2005, May 2006.
Copyright: Stephen G. Lenk (C) 2005, 2006. All rights reserved. 

This program is free software; you can redistribute it and/or  
modify it under the same terms as Perl itself.
Licenced under the Artistic Licence.

############################################################################
# This software comes with no guarantee of usefulness.
# Use at your own risk.
# There is no guarantee this code will find the best solution, or even any 
# solution, or that the solutions it finds will be correct or useful to 
# you.
#
# Use at your risk. Check any solutions you obtain.
#
# Stephen G. Lenk assumes no responsibility for the use of this software.
############################################################################

Primer3 is called by this code to verify that the PCR primers are OK.
Primer3 is Copyright (c) 1996,1997,1998,1999,2000,2001,2004
Whitehead Institute for Biomedical Research. All rights reserved.

2. Installation

MCPrimers.pm     - Place into Bio/MCPrimers.pm
CloningVector.pm - Place into Bio/Data/Plasmid/CloningVector.pm
mcprimers.pl     - Front end script. Make available to users.

PRIMER3DIR -   Set this environment variable to point to diectory containing
               Primer3 executable.

If PRIMER3DIR is not set, MCPrimers will set it to '.' by default.

MSWindows -    use primer3.exe
Other OS  -    use primer3_core

The code can be run in the download directory if an installation is not
desired.

3. Purpose

Creates molecular cloning PCR primer pairs for a given gene so that the
gene can be directionally inserted into a vector. pET-32a is the only
vector currently supported. Solver is generic, restriction enzymes and 
their order in the vector are specified in mcprimer.pl and in 
CloningVector.pm

4. Use

Use:     perl mcprimers.pl [options] -vector name dna.fasta > result.pr3
Options: [-h] [-shift search_shift] [-clamp (both | 3prime)]

-h        help
-vector   cloning vector name (i.e. pET-32a)
-shift    integer value to shift search of left RE match into gene
          use -shift 12 (or other value) if initial search fails
-clamp    GC clamp
          'both' will clamp last NT at both ends to G or C (default)
          '3prime' will clamp last NT on 3' ends to G or C
          
Example:  

perl mcprimers.pl -vector pET-32a -shift 12 -clamp 3prime cysS.fa > cysS.pr3

Note:        Use perl -Ilib if the modules are still in the local lib dir.

Limitations: Does not account for redundancy codes in FASTA files.

Note:        These runs use intermediate files. This means that
             CGI or other use that can overwrite these files is
             a bad idea. 

5. Test Results (V2.1)

Passes test.pl clean on Windows XP with Activestate Perl 5.8.7

Passes test.pl clean on OSX with Activestate Perl 5.8.8

Passes test.pl clean on Linux with the standard installed Perl 5.8.7 

Note that HIcysS.fa file needs to be written out in the right text format
prior to running test.pl

OSX with non-Activestate Perl (5.8.1) there are different results each time 
from mcprimers !?

---------------

Enjoy,

Steve Lenk
slenk@emich.edu
May 2006


Changes
Bio::MCPrimers package change log

V1.03 
Something like initial release.

V1.04 
Was on CPAN for maybe 6 months.

V1.5.0
Added seperate data access / fixed in-line issue.

V1.5.1
Package updated to better reflect good CPAN release practices.
No redundant solutions.
test.pl added

V2.0
Trying to appease CPAN version rules. 

V2.1
Documentation updates
Added no args test case to mcprimers.pl (help message)
Removed shebang from mcprimers.pl

Added system call check in test.pl


test.pl
# test.pl - test Bio::MCPrimers package
$VERSION='2.1'; 

# Author: Steve Lenk, May 2006

# Copyright (C) Stephen G. Lenk 2006.
# Artistic License applied.

# Use:    perl test.pl
# Output: OK or Not OK
# Return: 0 = OK, 1 = Not OK

use strict;
use warnings;

# use HIcysS.fa to generate a new test file
my $test = 1;
print "Test $test - ";
my $status = system('perl -Ilib mcprimers.pl -vector pET-32a -clamp 3prime -shift 24 HIcysS.fa > generated_file.pr3');
if ($status != 0) {
    die "\n\nError in system call to mcprimers.pl - is primer3 available?\n\n";
}

my $good_file;
my $generated_file;

open $good_file,      '<HIcysS.pr3'         or die "Can\'t open HIcysS.pr3\n";
open $generated_file, '<generated_file.pr3' or die "Can\'t open generated_file.pr3\n";

my $good_line      = <$good_file>;
my $generated_line = <$generated_file>;

# OK so far
$status = 1;

my %good;
my %generated;
my $line = 0;

# compare good and generated solutions
CHECK: while (defined $good_line) {

    chomp $good_line;
    chomp $generated_line;

    # retain right primers in hashes
    if ($good_line =~ /RIGHT PRIMER(.*)/) {
        $good{$1} = 1;
    }    
    if ($generated_line =~ /RIGHT PRIMER(.*)/) {
        $generated{$1} = 1;
    }
    
    # compared line-by-line
    $line = $line + 1;
    if ($good_line ne $generated_line) {
        $status = 0;
        print "Difference at line $line\n";
    }

    $good_line      = <$good_file>;
    $generated_line = <$generated_file>;
}

close $good_file;
close $generated_file;

# see if good generated any unique solutions
foreach (keys %good) {
    if (not defined $generated{$_}) {
        $status = 0;
        print "Unique good: $_\n";
    }
}

# see if generated generated any unique solutions
foreach (keys %generated) {
    if (not defined $good{$_}) {
        $status = 0;
        print "Unique generated: $_\n";
    }
}

# see if there are redundant solutions in good
my %regenerated1;
foreach (keys %good) {
    if (defined $regenerated1{$_}) {
        $status = 0;
        print "Regenerated in good: $_\n";
    }
    $regenerated1{$_} = 1;
}

# see if there are redundant solutions in generated
my %regenerated2;
foreach (keys %generated) {
    if (defined $regenerated2{$_}) {
        $status = 0;
        print "Regenerated in generated: $_\n";
    }
    $regenerated2{$_} = 1;
}

# clean up
unlink 'generated_file.pr3';

#simple presentation of results
if ($status == 1) {
    print "OK\n";
    exit 0;
}
else {
    print "Not OK\n";
    exit 1;
}


lib/Bio/Data/Plasmid/CloningVector.pm
package Bio::Data::Plasmid::CloningVector; 
our $VERSION = '1.00';

# Stephen G. Lenk (C) 2006. All rights reserved. 

# This program is free software; you can redistribute it and/or 
# modify it under the same terms as Perl itself.

# Licenced under the Artistic Licence.

# This software comes with no guarantee of usefulness. 
# Use at your own risk. Check any solutions you obtain. 
# Stephen G. Lenk assumes no responsibility for the use of this software.


# Use: cloning_vector_data ( $vector_name,  # vector name (i.e. pET-32a)
#                            $re_ra,        # restriction enzyme pattern
#                            $re_name_rh,   # names of restriction enzymes
#                            $ecut_loc_ra,  # cut location in enzyme
#                            $vcut_loc_ra ) # cut location in vector

# Returns: 0=fails (vector not found), 1=success

 
use strict;
use warnings;

sub cloning_vector_data {

    my ( $vector_name,  # vector name (i.e. pET-32a)
         $re_ra,        # restriction enzymes pattern
         $re_name_rh,   # names of restriction enzymes
         $ecut_loc_ra,  # cut location in enzyme
         $vcut_loc_ra   # cut location in vector
       ) = @_;  


    # check name of cloning vector
    if ($vector_name eq 'pET-32a') {

        push @{$re_ra}, 'TTCGAA';      # NspV       (TT^CGAA) (labrat site) 
        push @{$re_ra}, 'AGATCT';      # BglII      (A^GATCT)
        push @{$re_ra}, 'GGTACC';      # KpnI       (GGTAC^C)
        push @{$re_ra}, 'CCATGG';      # NcoI       (C^CATGG)
        push @{$re_ra}, 'GATATC';      # EcoRV      (GAT^ATC)
        push @{$re_ra}, 'GGATCC';      # BamHI      (G^GATCC)
        push @{$re_ra}, 'GAATTC';      # EcoRI      (G^AATTC)
        push @{$re_ra}, 'GAGCTC';      # SacI       (GAGCT^C)
        push @{$re_ra}, 'GTCGAC';      # SalI       (G^TCGAC)
        push @{$re_ra}, 'AAGCTT';      # HindIII    (A^AGCTT)
        push @{$re_ra}, 'GCGGCCGC';    # NotI, EagI (GC^GGCCGC)
        push @{$re_ra}, 'CTCGAG';      # AvaI, XhoI (C^TCGAG)   
    
        $re_name_rh->{TTCGAA}   = 'NspV';
        $re_name_rh->{AGATCT}   = 'BglII';
        $re_name_rh->{GGTACC}   = 'KpnI';
        $re_name_rh->{CCATGG}   = 'NcoI';
        $re_name_rh->{GATATC}   = 'EcoRV';
        $re_name_rh->{GGATCC}   = 'BamHI';
        $re_name_rh->{GAATTC}   = 'EcoRI';
        $re_name_rh->{GAGCTC}   = 'SacI';
        $re_name_rh->{GTCGAC}   = 'SalI';
        $re_name_rh->{AAGCTT}   = 'HindIII';
        $re_name_rh->{GCGGCCGC} = 'NotI, EagI';
        $re_name_rh->{CTCGAG}   = 'AvaI, XhoI'; 
    
        push @{$ecut_loc_ra}, 2;
        push @{$ecut_loc_ra}, 1;
        push @{$ecut_loc_ra}, 5;
        push @{$ecut_loc_ra}, 1;
        push @{$ecut_loc_ra}, 3;
        push @{$ecut_loc_ra}, 1;
        push @{$ecut_loc_ra}, 1;
        push @{$ecut_loc_ra}, 5;
        push @{$ecut_loc_ra}, 1;
        push @{$ecut_loc_ra}, 1;
        push @{$ecut_loc_ra}, 2;
        push @{$ecut_loc_ra}, 1;    
    
        push @{$vcut_loc_ra}, 2;
        push @{$vcut_loc_ra}, 0;
        push @{$vcut_loc_ra}, 2;
        push @{$vcut_loc_ra}, 2;
        push @{$vcut_loc_ra}, 0;
        push @{$vcut_loc_ra}, 1;
        push @{$vcut_loc_ra}, 1;
        push @{$vcut_loc_ra}, 2;
        push @{$vcut_loc_ra}, 2;
        push @{$vcut_loc_ra}, 2;
        push @{$vcut_loc_ra}, 0;
        push @{$vcut_loc_ra}, 2;  

        return 1;

    }
    else {

        return 0;
    }   
}

return 1;


mcprimers.pl
$VERSION = '2.1'; 

# mcprimers.pl - Designs molecular cloning PCR primers.

# Author:    Stephen G. Lenk, November 2005, May 2006.
# Copyright: Stephen G. Lenk (C) 2005, 2006. All rights reserved. 

# This program is free software; you can redistribute it and/or  
# modify it under the same terms as Perl itself.
# Licenced under the Artistic Licence.

#########################################################################
# This software comes with no guarantee of usefulness. 
# Use at your own risk. Check any solutions you obtain.
# Stephen G. Lenk assumes no responsibility for the use of this software.
#########################################################################

# Limitations: Does not account for redundancy codes in FASTA files.

# Note:        These runs use intermediate files. This means that
#              CGI or other use that can overwrite these files is
#              a bad idea. 

# see Bio::MCPrimers for POD

# Use with V2.1 of BIO::MCPrimers.pm

use strict;
use warnings;

use Bio::MCPrimers;


my $input_fh;
my $use_msg = qq/
MCPrimers generates molecular cloning PCR primers.
Copyright (C) 2005, 2006 Stephen G. Lenk.

Use:     perl mcprimers.pl [options] -vector name dna.fasta > result.pr3
Options: [-h] [-shift search_shift] [-clamp (both | 3prime)]

-h        help
-vector   cloning vector (i.e. pET-32a)
-shift    integer value to shift search of left RE match into gene
          use -shift 24 (or other value) if initial search fails
-clamp    GC clamp
          'both' will clamp last NT at both ends to G or C (default)
          '3prime' will clamp last NT on only 3' ends to G or C

Use at your risk. Check any solutions you obtain.
          
/;


my %flag;                       # hash for all flags
$flag{clamp}        = 'both';
$flag{search_shift} = 0;

my $flag;                       # flag currently being checked
my $vector_name;                # name of cloning vector

# trap no args case
if (@ARGV == 0) {
    
    # help
    print STDERR $use_msg;
    exit(1);
}

# command line options
while (@ARGV > 0) { 

    $flag = shift @ARGV;
    my $found = 0;

    if ($flag eq '-h') {
    
        # help
        print STDERR $use_msg;
        exit(1);
    }     
    elsif ($flag eq '-vector') {
        
        # select cloning vector
        if (@ARGV) {
            $vector_name = shift @ARGV;
            if (defined $vector_name and $vector_name ne '') {
                $found = 1;
            }
        }
    }   
    elsif ($flag eq '-shift') {
    
        # search shift
        if (@ARGV) {
            $flag{search_shift} = shift @ARGV;
            if ($flag{search_shift} =~ /^(\d)+$/) {
                $found = 1;
            } else {
                die "\nError: -shift $flag{search_shift} not recognised\n\n";
            }
        } else {
            die "\nError: -s needs an argument\n\n";
        }
    }
    elsif ($flag eq '-clamp') {

        # GC clamping
        if (@ARGV) {
            $flag{clamp} = shift @ARGV;
            if ($flag{clamp} eq 'both' or $flag{clamp} eq '3prime') {
                $found = 1;
            } else {
                die "\nError: -clamp $flag{clamp} not recognised\n\n";
            }
        } else {
            die "\n-Error: clamp needs an argument\n\n";
        }
    }
    
    # unknown flag
    if (not $found and @ARGV) {
        die "\nError: Option $flag not recognised\n\n";
    }
    
    # Should be a FASTA file name left at the end
    if ($found and @ARGV == 0) {
        die "\nError: Specify a FASTA file for input\n\n";  
    }
}


# Check that cloning vector has been defined
if (not defined $vector_name or $vector_name eq '') {
    die "\n\nError: Vector name not defined\n\n";
}


# open input file
my $infile = $flag;
if (not defined $infile) { 
    print STDERR $use_msg;
    exit(0);
}
open $input_fh, "<$infile" or die "\n\nError: Can\'t use $infile for input\n\n";


my $line;               # read lines
my $gene = '';          # then load ONLY nucleotides into $gene

# read fasta file

# toss > annotation
$line = '>';
while (substr($line,0,1) eq '>') {
    $line = <$input_fh>;   
}

# gete sequence data
while (defined $line) {

   chomp $line;
   $gene .= $line;
   $line = <$input_fh>;
}


my @re;           # restriction enzymes pattern
my %re_name;      # names of restriction enzymes
my @ecut_loc;     # cut location in enzyme
my @vcut_loc;     # cut location in vector

use Bio::Data::Plasmid::CloningVector;  # details of the plasmid used as a vector
my $status = Bio::Data::Plasmid::CloningVector::cloning_vector_data($vector_name,
                                                                    \@re, 
                                                                    \%re_name, 
                                                                    \@ecut_loc, 
                                                                    \@vcut_loc);            
if ($status == 0) {
    die "\n\nError: Data not found for cloning vector $vector_name\n\n";
}

my $version;      # version of MCPrimers returned from MCPrimers

# invoke solver
my $answer_ar;
$answer_ar = Bio::MCPrimers::find_mc_primers($gene, 
                                             \%flag, 
                                             \$version, 
                                             \@ecut_loc,
                                             \@vcut_loc,
                                             @re);


# Copyright notices, Primer3 here as MCPrimers uses Primer3
my $copr = 
qq/
|------------------------------------------------------------------|
| MCPrimers V$version                                                   |    
| Copyright (c) 2005,2006 Stephen G. Lenk. All rights reserved     |
| Primer3 Copyright (c) 1996,1997,1998,1999,2000,2001,2004         |  
| Whitehead Institute for Biomedical Research. All rights reserved |
|------------------------------------------------------------------|

Cloning vector    = $vector_name
Clamp flag        = $flag{clamp}
Shift flag        = $flag{search_shift}
Original sequence =
/;

print $copr;

# give them original gene sequence for reference
while ($gene =~ /(.{1,60})/g) {
    print "$1\n"; 
}
print "\n\n";

if (not defined $answer_ar or @{$answer_ar} == 0) { 

    # No solution found
    print "\nNo solution found\n\n";

} else {

    # Dump the solutions
    my $count = 0;
    foreach my $answer_hr (@{$answer_ar}) {

        # handle count
        $count += 1;
        my $count_text = 
            "=========\n=========  Solution # $count\n=========";
        
        # compose result text
        my $result = qq/
Start codon at  $answer_hr->{start}
Stop codon  at  $answer_hr->{stop}
Left RE site  = $re_name{$answer_hr->{left_re}} ($answer_hr->{left_re})
Right RE site = $re_name{$answer_hr->{right_re}} ($answer_hr->{right_re})

Primer3 analysis of PCR primers designed by MCPrimers:

$answer_hr->{primer3}/;

        print "$count_text\n$result\n\n";
    }
}


exit(0);
